### ISSUE TITLE
Validation during training

### ISSUE THREAD

**Issue by arjungpt28** [bytedance/video-SALMONN-2#17]
Hi,

Thank you for releasing this great work! I'm in the process of fine-tuning the model via SFT and had some questions about the intended data format and the behavior of the validation loss. I'm noticing something strange, which I want to ask whether this is expected or not. Firstly, this is the format of my dataset (one for each type of the three questions in LLaVA-Video-178K):

Captioning:
```
{
        "video": "data/LLaVA-Video-178K/videos/ytb_3TZdidJx57k.mp4",
        "audio": "data/LLaVA-Video-178K/audio/ytb_3TZdidJx57k.wav",
        "conversations": [
            {
                "from": "human",
                "value": "<image>\nNarrate the video's unfolding events in a descriptive manner."
            },
            {
                "from": "gpt",
                "value": "The video begins with an exterior shot of a building labeled 'MEDICAL LIBRARY.' The building, adorned with ivy and featuring a staircase flanked by columns, transitions to the interior where a person in a white lab coat is seated at a desk, writing or taking notes. The library's bookshelves are labeled 'CIRCULATORY A-L' and 'CARDIOLOGY A-M.' The person checks their watch and continues writing before turning to speak while holding a pen.\n\nThe scene shifts to an operating room where a person in surgical attire performs surgery, with another similarly dressed person standing near a table of medical instruments. The video then transitions to a man in a brown suit holding a microphone, speaking in front of an audience with blue curtains and stage lights in the background. He holds a piece of paper while presenting. The scene changes to two men on a stage with a cash register between them; one has white hair and a beard, wearing a red jacket, and the other wears glasses, a bow tie, and a vest. A person with a tall, flat-top hairstyle is seated next to the man with white hair, with a palm tree visible in the background.\n\nThe narrative continues with two men seated on a stage with blue curtains. The man on the left has a tall, flat-top hairstyle and wears a light blue shirt with a purple tie, with a broom standing upright next to him. The man on the right, in a brown suit, has a unique tufted hairstyle. The scene transitions to a close-up of the man in the brown suit seated next to a carved pumpkin. Another man with a short, dark hairstyle, wearing a white shirt and black tie, is seated next to the pumpkin, with a large silver coffee pot on a table beside him. The video returns to the medical library, where a man in a white lab coat reacts with surprise or concern while looking at a computer monitor. The clip ends with the man in the brown suit being interviewed by another man in a brown suit holding a microphone.\n\nThe final scene shows the man with the tufted hairstyle on a stage with blue curtains, being interviewed by a person in a brown suit holding a microphone. The man with the tufted hairstyle speaks animatedly, displaying various expressions from surprise to concern, with a carved pumpkin in the background adding a festive touch. The video concludes with the man looking to the side while the interviewer continues to hold the microphone towards him."
            }
        ]
    },
```

Open-Ended:
```
{
        "video": "data/LLaVA-Video-178K/videos/ytb_3btUGmROy9k.mp4",
        "audio": "data/LLaVA-Video-178K/audio/ytb_3btUGmROy9k.wav",
        "conversations": [
            {
                "from": "human",
                "value": "<image>\nWhat is the initial visual element in the video?"
            },
            {
                "from": "gpt",
                "value": "The video begins with a blurred world map with various regions highlighted in different colors."
            },
            {
                "from": "human",
                "value": "What is the main topic of the video?"
            },
            {
                "from": "gpt",
                "value": "The main topic of the video is 'What's the Most Spoken Language'."
            },
            {
                "from": "human",
                "value": "Which language is indicated as the most spoken language in the video?"
            },
            {
                "from": "gpt",
                "value": "English is indicated as the most spoken language."
            },
            {
                "from": "human",
                "value": "Which country's map is shown with a large Union Jack flag overlay?"
            },
            {
                "from": "gpt",
                "value": "The map of the United Kingdom is shown with a large Union Jack flag overlay."
            },
            {
                "from": "human",
                "value": "How many native English speakers are there according to the video?"
            },
            {
                "from": "gpt",
                "value": "There are 370 million native English speakers."
            }
        ]
    },
```

MCQ:
```
{
        "video": "data/LLaVA-Video-178K/videos/ytb_XNvOh3wINU0.mp4",
        "audio": "data/LLaVA-Video-178K/audio/ytb_XNvOh3wINU0.wav",
        "conversations": [
            {
                "from": "human",
                "value": "<image>\nWhat was the jersey number of the player in a white uniform who jogged towards the camera?\nA. 30\nB. 12\nC. 23\nD. 45\nPlease provide your answer by stating the letter followed by the full option."
            },
            {
                "from": "gpt",
                "value": "B. 12."
            },
            {
                "from": "human",
                "value": "What advertisement was displayed in the background during the final sequence?\nA. Pepsi\nB. State Farm\nC. Nike\nD. Adidas\nPlease respond with only the letter of the correct answer."
            },
            {
                "from": "gpt",
                "value": "B."
            }
        ]
    },
```

Could you please confirm for me whether this is correct (particularly for the open-ended and MCQ style of questions, which can have multiple questions for the same video)?

Next, when I added a `--evaluation_strategy` flag to do validation, I observed the initial validation loss was high, even when initializing with your final model. To observe what was happening, I tried adding a `--do_test flag` to see the output of the model, and this is what I see:

```
{
        "id": [
            "data/LLaVA-Video-178K/videos/ytb_Peq6FprQImg.mp4",
            "data/LLaVA-Video-178K/audio/ytb_Peq6FprQImg.wav"
        ],
        "prompt": [
            {
                "from": "human",
                "value": "<image>\nNarrate the video's unfolding events in a descriptive manner."
            }
        ],
        "pred": "The video takes place in a store, likely a supermarket or department store, characterized by shelves filled with various products. It opens with a scene where a woman with long blonde hair, dressed in a blue vest and white shorts, is diligently organizing items on a lower shelf. In the background, a man holds two watermelons while standing near a shopping cart filled with items. This scene captures a casual moment as everyday activities unfold. Text appears on the screen saying 'Damn shawty' followed by 'those are some nice melons!' (highlighting someone's admiration for the watermelons). The camera then focuses more closely on the woman, who is now interacting with a man wearing a white t-shirt and jeans. During their interaction, the woman remarks, \"Those are some nice melons!\" to which the man responds with an apologetic 'I'm sorry.' However, she immediately points out, \"But that's really disrespectful,\" indicating that his comment was inappropriate. She warns him, \"And if you say that to me again, I'm gonna have to kick you out,\" asserting her authority in maintaining the store's decorum.\n\nAs the interaction continues, the woman is seen gesturing with her hands while talking to the same man, who stands nearby holding one of the watermelons. The man in the background still clutches the other watermelon. Subtitles at the bottom of the screen read 'I'm talking about' followed by 'my friend has nice' and 'wasn't talking about you,' suggesting the man's attempt to explain his comment. Despite his apology, 'I'm sorry' accompanied by a sad emoji, the woman firmly states 'No' and adds, 'Nah you're good,' with a laughing emoji, indicating that she is jesting and not actually upset. This part of the clip shows the woman returning to organize items on the shelf while the man and the man with the watermelons stand nearby, with the latter still holding the watermelons. Shelves stocked with boxes and other merchandise form the backdrop, along with a few other shoppers going about their business, adding to the lively yet orderly atmosphere of the store.\n\nThis sequence effectively portrays a minor yet relatable misunderstanding that is resolved amicably, highlighting the woman's assertiveness in upholding store etiquette while also demonstrating her willingness to forgive and laugh, thus maintaining a light-hearted and realistic depiction of a situation that could occur in any public setting.\n\n<|im_end|>",
        "ref": "\n\n"
    },
```

While the prediction looks quite good, weirdly, it seems like some of the references are emptied out, by this code snippet in `llava/dataset/av_dataset.py`:

```
if self.is_test:
    if "prefix" not in sources[0]["conversations"][-1]:
        sources[0]["conversations"][-1]["value"] = ""
```

At the same time, the references for the open-ended and MCQ style questions seem off, too, which may also explain the high validation loss:

```
{
        "id": [
            "data/LLaVA-Video-178K/videos/ytb_oRC7379MoxI.mp4",
            "data/LLaVA-Video-178K/audio/ytb_oRC7379MoxI.wav"
        ],
        "prompt": [
            {
                "from": "human",
                "value": "<image>\nDoes the player in white and yellow uniform make a successful shot after dribbling the ball?"
            },
            {
                "from": "gpt",
                "value": "No, the player in white and yellow uniform loses his balance and falls to the floor instead of making a shot."
            },
            {
                "from": "human",
                "value": "What sequence of events occurs after the player in white and yellow uniform loses his balance?"
            },
            {
                "from": "gpt",
                "value": "After the player in white and yellow uniform loses his balance, he falls to the floor and is surrounded by players in green and black uniforms, specifically numbers 7 (Brown) and 12 (Williams). The focus then shifts to a close-up of a man in the audience engaged in a lively conversation."
            },
            {
                "from": "human",
                "value": "What does the camera focus on after the player in white and yellow uniform falls?"
            }
        ],
        "pred": "The camera focuses on a close-up of a man in the audience wearing a white hoodie over a blue shirt with a graphic print, and a chain necklace.<|im_end|>",
        "ref": "\n\nNo, the player in white and yellow uniform loses his balance and falls to the floor instead of making a shot.\n\nAfter the player in white and yellow uniform loses his balance, he falls to the floor and is surrounded by players in green and black uniforms, specifically numbers 7 (Brown) and 12 (Williams). The focus then shifts to a close-up of a man in the audience engaged in a lively conversation.\n\n"
    },
```
```
{
        "id": [
            "data/LLaVA-Video-178K/videos/ytb_wYiv1YbVxOQ.mp4",
            "data/LLaVA-Video-178K/audio/ytb_wYiv1YbVxOQ.wav"
        ],
        "prompt": [
            {
                "from": "human",
                "value": "<image>\nWhat kind of headband does the person wear with the white robe?\nA. A blue headband with stars\nB. A pink headband with pom-poms\nC. A red headband with stripes\nD. A headband adorned with colorful flowers\nPlease provide your answer by stating the letter followed by the full option."
            },
            {
                "from": "gpt",
                "value": "B. A pink headband with pom-poms."
            },
            {
                "from": "human",
                "value": "What kind of light is used to illuminate the room?\nA. A green LED strip light and a small lamp\nB. A chandelier and a floor lamp\nC. A string of fairy lights\nD. A ceiling fan with lights\nPlease respond with only the letter of the correct answer."
            }
        ],
        "pred": "A.<|im_end|>",
        "ref": "\n\nB. A pink headband with pom-poms.\n\n"
    },
```

Can you please help me understand what might be going on and whether my understanding is correct? Any guidance you could provide would be greatly appreciated. Thank you so much!

**Comment by TCL606** [comment]
The data format is correct. It's strange that the ref is not conversations[-1]["value"]. I want to confirm whether you have modified the code of the dataset. Normally, during testing, the existing dataset will treat all previous conversations (conversations[:-1]) as prompts, and the ref should strictly be conversations[-1]["value"].
